vpr/src/place/analytic_placer.h

```diff
@@ -0,0 +1,160 @@
+#ifndef VPR_ANALYTIC_PLACEMENT_H
+#define VPR_ANALYTIC_PLACEMENT_H
+
+#include "vpr_context.h"
+#include "timing_place.h"
+#include "PlacementDelayCalculator.h"
+
+typedef signed long int wirelen_t;
+
+template<typename T>
+struct EquationSystem;
+
+class AnalyticPlacer {
+  public:
+    AnalyticPlacer(VprContext& ctx);
+
+    void ap_place();
+
+  private:
+    // for CutSpreader to access legal_pos, block_locs, etc.
+    friend class CutSpreader;
+
+    // TODO: PlacerHeapCfg should be externally configured & supplied
+    struct PlacerHeapCfg {
+        float alpha, beta;
+        int criticalityExponent;
+        int timingWeight;
+        float solverTolerance;
+        bool placeAllAtOnce;
+        int spread_scale_x, spread_scale_y;
+    };
+
+    PlacerHeapCfg tmpCfg;
+
+    int max_x = 0, max_y = 0; //size of the grid
+    VprContext& vpr_ctx;
+    const DeviceContext& device_ctx;
+    const ClusteredNetlist& clb_nlist;
+    PlacementContext& place_ctx;
+
+    std::vector<std::vector<int>> num_legal_pos; /* [0..num_sub_tiles - 1][0..num_legal_pos-1] */
+    std::vector<std::vector<std::vector<t_pl_loc>>> legal_pos;
+    std::vector<std::vector<std::vector<std::vector<std::vector<t_pl_loc>>>>> fast_tiles;
+
+    // marker for blks not solved in current equation
+    int32_t dont_solve = std::numeric_limits<int32_t>::max();
+    // marker for blks not in any macros
+    size_t not_in_macros = std::numeric_limits<size_t>::max();
+
+    struct MacroInfo;
+
+    struct BlockInfo {
```

Comment:
Need comments for all these data members.  Some seem to be copies of what is in the clb netlist; better not to copy those I think (unless you are getting ready to handle netlists other than the clb netlist -- i.e. unclustered / atom netlist).  If you're decoupling from a specific netlist for that reason it would make sense, but you would need to comment that so your reasoning is clear.